[BE][NAS] 야구게임 API구현 및 OAuth로그인 기능 구현

Malloc72P opened the pull request:
안녕하세요. 백엔드의 NAS라고 합니다
야구게임 기능구현 및 OAuth 적용을 완료해서 PR을 남기게 되었습니다.

# 야구게임 구현
* 야구는 테스트코드를 작성해가면서 기능구현을 했습니다
* 그렇긴 한데, 아직 테스트코드를 작성하는게 많이 미숙한 듯 합니다
![image](https://user-images.githubusercontent.com/41663269/118095911-a91fb700-b40b-11eb-93c2-5e15b2f3d13f.png)

# 데이터베이스 설계
* 프로젝트의 ERD는 다음과 같습니다
![image](https://user-images.githubusercontent.com/41663269/118231130-9ec3f280-b4c9-11eb-8d4c-57d7dd17b34f.png)
### 어그리게이트
* Game과 Team을 각각 분리된 어그리게이트로 만들었습니다.
* Game을 어그리게이트 루트로 삼아서 게임에 관련된 모든 정보를 관리하도록 했습니다
* Game과 별개로 팀 자체의 정보는 Team을 어그리게이트 루트로 삼아서 관리하도록 설계했습니다.
* Game을 만들게되면 Team의 정보를 이용해서 팀이 게임에 참여했다는 정보를 저장하도록 했습니다.(TeamParticipatingInGame)
* 이렇게 해서 게임마다 선수의 타순을 자유롭게 설정할 수 있도록 설계했습니다.
### 점수 관리
* 게임이 이닝을 리스트로 가지도록 설계했습니다
* 게임에서 홈팀이나 원정팀의 총점을 얻으려면, 이닝별 모든 점수를 더해서 각 팀의 총점을 구할 수 있도록 했습니다.
### 선수 기록관리
* 팀이 게임에 참여하게 되면, 선수가 게임에 참여했다는 정보도 저장하도록 설계했습니다(PlayerParticipatingInGame)
* 이렇게 해서 게임마다 선수의 정보(타순, 타석, 안타, 아웃, 포지션)를 관리할 수 있도록 했습니다.

# OAuth 로그인 기능 구현
### 사용자 프로필정보 요청
* OAuth 로그인은 구글 OAuth를 통해 구현하였습니다
* 사용자가 로그인을 완료하고, Authorization Code가 CallbackURL에 매칭되는 컨트롤러 메서드로 넘어오면, 코드를 가지고 액세스 토큰을 요청하도록 구현했습니다
* 액세스 토큰을 획득하면, 토큰으로 사용자의 프로필 정보를 불러와서 디비에 저장하도록 구현했습니다.
### JWT 토큰 발급
* 사용자의 프로필 정보를 불러와서 저장한 다음에는 저희 서비스에 대한 액세스토큰과 리프래시 JWT토큰을 발급하도록 구현했습니다
* 아래와 같은 형태로 오도록 구현했습니다
![image](https://user-images.githubusercontent.com/41663269/118113637-7da7c700-b421-11eb-9551-c067d88de6f3.png)

### 토큰 리프래시
* 액세스토큰이 만료되면 리프래시 토큰으로 토큰 리프래시 요청을 해서 새로운 액세스 토큰과 리프래시 토큰을 받아오도록 구현했습니다
![image](https://user-images.githubusercontent.com/41663269/118113903-d5dec900-b421-11eb-96b0-27ecea3c99e8.png)

### 토큰 검사
* 토큰은 인터셉터를 이용해서 검사하도록 구현했습니다.
* `@Auth`어노테이션 유뮤를 가지고 검사할지 말지를 결정할 수 있도록 구현했습니다
```java
@Auth
@PatchMapping("/{gameId}")
public void joinIn(@PathVariable("gameId") int gameId, HttpServletRequest request) {/*...*/}
```

# 야구 게임 API
### 게임목록조회 API
![image](https://user-images.githubusercontent.com/41663269/118113983-f3139780-b421-11eb-9578-bc90e1a03f0f.png)

###  게임 참여하기 API
* 게임에 참여요청을 합니다
* Request Header의 Authorization에 담긴 액세스토큰을 가지고 어떤 사용자가 요청한건지 검사하도록 구현했습니다
![image](https://user-images.githubusercontent.com/41663269/118114118-21917280-b422-11eb-9dc2-78aa195a92b2.png)
* 만약 토큰이 유효하지 않으면 , 게임 참여하기는 실패하고, 다음과 같은 형식으로 응답해주도록 구현했습니다.
![image](https://user-images.githubusercontent.com/41663269/118114374-7503c080-b422-11eb-9321-afefd9c63e60.png)
* 또, 누군가에 의해 이미 선점된 게임이라면, 참여요청은 실패하고, 아래와 같은 응답을 받도록 구현했습니다
![image](https://user-images.githubusercontent.com/41663269/118114660-da57b180-b422-11eb-8187-962f37884846.png)

### 게임 상세조회 API
* 게임에 대한 세부정보를 조회합니다
* 해당 API도 토큰을 검증하도록 구현했습니다

### 피치하기 API
* 투수가 피치하게끔 요청하는 API입니다
* 해당 API도 마찬가지로 토큰검증을 하도록 구현했습니다.
* 추가로, 아무도 참여하지 않은 게임에다가 피치 요청을 하면 실패하도록 구현했습니다
![image](https://user-images.githubusercontent.com/41663269/118115011-5225dc00-b423-11eb-841b-316b16ade154.png)
* 또한, 다른 사용자가 선점한 게임에 피치요청을 하면 실패하도록 구현했습니다
![image](https://user-images.githubusercontent.com/41663269/118115052-64077f00-b423-11eb-975d-8a1c13097d91.png)
* 오직 자신이 선점한 게임에 한해서만 피치요청이 성공하게끔 구현했습니다
![image](https://user-images.githubusercontent.com/41663269/118115097-784b7c00-b423-11eb-904a-2584456ac48d.png)

# 배포
### 배포는 다음과 같은 구조로 했습니다
![image](https://user-images.githubusercontent.com/41663269/118116647-7be00280-b425-11eb-8c91-d7c99ad6f61e.png)

이상으로 글을 마치겠습니다.
이번 리뷰도 잘 부탁드립니다!

Seohyoung committed:
[#14] feat: CRA 및 컴포넌트 생성

 - CRA 생성
 - 기본 컴포넌트 생성

Malloc72P committed:
[#11] feat: 스프링 앱 초기화

- gitignore 내용 추가

Malloc72P committed:
[#12] feat : MySQL DB연동을 위한 설정 추가

Malloc72P committed:
[#12] feat : DB 스키마 스크립트 작성

Malloc72P committed:
[#12] feat : 앱 실행했을 때 schema파일을 실행하도록 설정 추가

Malloc72P committed:
Merge pull request #15 from Malloc72P/BE/feature/#12/DataJdbcDependency

Be/feature/#12/data jdbc dependency : Spring Data JDBC의존성 추가 및 스키마 작업

skawnkk committed:
[#14]Style: globalStyle,Theme설정

skawnkk committed:
Merge pull request #16 from Malloc72P/fe/feature/#14/style

[#14]Style: globalStyle,Theme설정

Malloc72P committed:
[#2] feat : 팀 엔티티 생성 테스트

- 해당 구현을 위해 팀 엔티티 생성함
- 엔티티를 이용해서 데이터가 잘 생성되는지 확인하기 위해서 레포지토리 생성하고 테스트 클래스 작성함

Malloc72P committed:
[#2] chore : 현재 시점에서 불필요한 팀 생성자 제거

Malloc72P committed:
[#2] feat : 불필요한 getter 제거

Malloc72P committed:
[#2] feat : 팀 엔티티 생성 테스트 보강

- 배열에 팀 이름을 3개 담고, 3번 생성시도하고 테스트하도록 보강함

Malloc72P committed:
[#2] feat : 플레이어를 팀에 추가하는 기능 구현

- 해당 기능 구현을 위해 플레이어 엔티티 추가함
- 테스트코드를 통해 팀에 플레이어를 Set으로 관리하는 기능을 테스트함

Malloc72P committed:
[#2] chore : 메서드 체이닝으로 붙인 코드를 개행시킴

skawnkk committed:
#1/Feat:게임목록조회 api요청

skawnkk committed:
Merge pull request #17 from Malloc72P/fe/feature/#1/fetch

#1/Feat:게임목록조회 api요청

Malloc72P committed:
[#2] refactor: 플레이어 추가 테스트기능 리팩토링

- 메서드를 추출함
- 반복문을 forEach와 메서드 레퍼런스로 대체함

Seohyoung committed:
[#4] feat : startPage UI 진행 중

- Start Page 메인에서 데이터 패치 처리
- Team List 컴포넌트 내 스타일 작업 중

Seohyoung committed:
Merge pull request #18 from Malloc72P/FE/markUp/#4/startPageUI

[#4] feat : startPage UI 진행 중

skawnkk committed:
[#10]Feat: gamePage이동을 위한 기본라우터설정

Malloc72P committed:
[#2] feat : Game엔티티 구현 및 생성 테스트 메서드 작성

skawnkk committed:
[#10]Feat: id값 router에 전달해주기

Malloc72P committed:
[#2] refactor: 테스트 메서드 리팩토링

- 웬만하면 스테틱 변수를 참조하도록 수정
- 테스트 메서드 추출

Malloc72P committed:
[#2] feat : 새로운 ERD에 대응해서 코드수정

- 게임에서 팀의 게임참여정보를 들고있는 방향으로 수정중

Malloc72P committed:
[#2] feat : 게임 엔티티에 참여중인 팀 정보 추가

- Set으로 팀을 관리하도록 수정함
- 기존에는 home, away에 따로 저장했더니, 쿼리에서 Left Outer Join을 해댔음
- 이로인해 findById를 하면 실제로는 게임은 하나인데, 4개가 검색되는 이상현상이 발생함
- Set으로 바꿔서 해결함

skawnkk committed:
Feat. 배포된 API로 경로 변경, useFetch 적용

skawnkk committed:
[#10]Feat:배포API로 재설정, useFetch Hook적용

skawnkk committed:
Merge pull request #20 from Malloc72P/fe/feature/#10/router

useFetch적용.

Seohyoung committed:
[#4] feat : startPage teamName hover 시 색상 변환

Seohyoung committed:
Merge branch 'dev-FE' into FE/markUp/#4/startPageUI

Seohyoung committed:
Merge pull request #21 from Malloc72P/FE/markUp/#4/startPageUI

[#4] feat : startPage teamName hover 시 색상 변환

Malloc72P committed:
[#2] feat : 게임에 참여중인 선수 정보를 입력하는 기능 구현

- 게임에 참여한 선수를 List로 관리하도록 구현
- 선수의 순서는 batOrder로 설정해서 구현을 편리하게 함
- 추가로 테스트코드를 세분화함

Malloc72P committed:
[#2] chore: intelliJ 자동 리포매팅 기능 수행

Seohyoung committed:
fix : 에러 수정 중

Seohyoung committed:
fix : fetch 에러 수정

Seohyoung committed:
fix : Route 에러 수정

Malloc72P committed:
[#2] feat : 경기 스키마 수정

- 1,2,3루정보를 담을 수 있도록 수정
- 엔티티에서 초기상태의 게임을 생성하면 1,2,3루를 비어있게 설정
- 해당 기능을 테스트하는 코드 작성

Malloc72P committed:
[#2] feat : 이닝 엔티티 추가

- 각 이닝 별 점수를 관리할 수 있도록 구현
- 초기상태의 이닝이 1이고, 점수는 0인지 검사하는 코드 추가

Malloc72P committed:
[#2] feat : 스트라이크 및 공수교대, 이닝 넘어가기 기능구현

- 삼진아웃 구현
- 3아웃시 공수교대, 또는 이닝 넘어가기 구현

Malloc72P committed:
[#2] feat : 볼 상황의 기능 구현

- 볼넷상황시 타자가 1루진출하도록 구현
- 1루주자는 2루로 진출하는 기능 구현
- 아직 2에서 3루로 가는거랑 backHome하는 기능은 테스트 못함
- 추가로 backHome했을때 점수추가하는 기능구현 필요

Seohyoung committed:
fix : redirect 오류 해결

skawnkk committed:
Fix. useFetch 수정

skawnkk committed:
Merge pull request #23 from Malloc72P/fe/fixError

useFetch _ useState 관리 상태 1개로 정리.

Malloc72P committed:
[#2] feat : 볼넷 상황의 테스트 추가

- 테스트코드에서 게임을 초기화 하지 않던 문제 해결
- 만루상태의 볼넷을 통해 주자가 백홈하는것을 테스트함

Malloc72P committed:
[#2] refactor: 중복코드 제거

Malloc72P committed:
[#2] feat : 볼넷으로 점수가 발생하는 기능 및 테스트 코드 구현

- 해당 기능 구현과정에서 공수교대시 1,2,3루 정보를 리셋하지 않는 버그 발견하고 수정함

Malloc72P committed:
[#2] feat : 안타 구현 및 테스트 코드 작성

Malloc72P committed:
[#2] feat : 홈런기능 구현 및 테스트 코드 작성

Malloc72P committed:
[#2] feat : 이닝 별 점수기록을 조회하는 기능 추가

Malloc72P committed:
[#2] feat : 게임 기록 엔티티 구현 및 기록저장기능 구현

- 아웃발생시 다음타자가 현재타자로 기록되는 문제를 발견함
- 그래서 피치결과판정 전에 기록하도록 수정함
- 그 과정에서 스트라이크 아웃상황에서는 다음타자로 바꾸지 않던 문제를 발견해서 해결함

Malloc72P committed:
[#2] feat : 게임 기록 테스트 추가구현

- 각각의 행동으로 획득한 점수도 기록하도록 수정함

Malloc72P committed:
[#2] feat : 각각의 행동을 하면 레파지토리에 game을 저장하도록 수정

Malloc72P committed:
[#2] feat : 선수의 기록을 저장할 수 있도록 기능구현

Malloc72P committed:
[#2] refactor: IntelliJ로 코드 자동 리포매팅

Malloc72P committed:
[#2] feat: 게임 종료기능 구현 및 테스트코드 추가

Malloc72P committed:
Merge pull request #37 from Malloc72P/BE/feature/#2/Entity

Be/feature/#2/entity 야구 게임 구현

Malloc72P committed:
[#36] feat: 커맨드라인 러너로 테스트 데이터 입력

Malloc72P committed:
[#36] feat: 게임 목록조회 API 구현

Malloc72P committed:
[#36] feat: 게임 참여하기 및 상세조회 API 구현

Malloc72P committed:
[#36] feat: 피치하기 API 구현

Malloc72P committed:
[#36] fix: 게임에 선수정보를 넣는 과정에서 두번 중복으로 넣는 오류 해결

Malloc72P committed:
[#36] feat: 난수테이블 확률조정

- 홈런과 안타가 너무 자주 터진다. 확률을 줄였다.
- 반면 스트라이크와 볼은 확률을 올렸다

Malloc72P committed:
[#36] feat: 이닝번호가 0이 아닌 1부터 시작하도록 수정

Malloc72P committed:
[#36] feat: 피치하기 API의 결과에 해당하는 DTO의 필드순서 변경

- 점수랑 게임상태를 먼저 보고 싶어서 이렇게 수정함

Malloc72P committed:
[#36] feat: 예외처리 컨트롤러 구현

- 못찾았거나 이미 가득찬 방에 들어가려는 경우, 지정된 에러데이터를 응답하도록 구현

Malloc72P committed:
[#36] feat: Method Not Allowed에 대한 예외처리 추가

Malloc72P committed:
[#36] chore: 코드 리포매팅 수행

Malloc72P committed:
[#36] feat: 테스트용으로 데이터를 리셋하는 기능 추가

Malloc72P committed:
[#36] feat: 시작하지 않은 게임에 대해 피치요청을 하면 에러를 응답하도록 구현

Malloc72P committed:
Merge pull request #47 from Malloc72P/BE/feature/#36/API

Be/feature/#36/api API 기능구현

Malloc72P committed:
[#48] chore: 패키지 분류

Malloc72P committed:
[#48] feat: OAuth 설정정보 파일 gitignore에 추가

Malloc72P committed:
[#48] chore: dto패키지 추가분류

Malloc72P committed:
[#48] feat: oauth 설정정보를 관리하는 서비스 생성

Malloc72P committed:
[#48] feat: 설정정보를 불러오면 출력되는 메세지를 static변수로 추출

Malloc72P committed:
[#48] feat: OAuth 로그인 기능 구현

- authorization code 획득 후, 이를 통해 액세스 토큰 획득
- 획득한 토큰으로 유저 정보 획득
- 유저정보를 디비에 저장 후, 브라우저에게 jwt토큰과 userid를 전송

Malloc72P committed:
[#48] chore: 좀 더 나은 이름으로 클래스 이름 수정 및 패키지 재분류

Malloc72P committed:
[#48] feat: Auth 어노테이션 추가

Malloc72P committed:
[#48] feat: Auth 어노테이션 추가 및 인터셉터 적용

- Auth 어노테이션을 부착한 컨트롤러에 한해서 토큰검증을 실시하도록 구현

Malloc72P committed:
[#48] chore: 코드 자동 리포매팅 수행

Malloc72P committed:
[#48] feat: 인터셉터 기능보강

- 인터셉터에서 토큰검사하고, 안에있는 유저아이디를 꺼내도록 구현
- 꺼낸 유저아이디를 HttpServletRequest객체를 이용해서 컨트롤러한테 전달할 수 있도록 구현

Malloc72P committed:
[#48] chore: 불필요한 import 제거

Malloc72P committed:
[#48] feat: 리프래시 토큰 구현

- 추가로 JWT 토큰정보를 보낼 때, 액세스, 리프래시 토큰의 만료시간도 주도록 구현

Malloc72P committed:
[#48] fix: 볼 카운트가 쌓이지 않는 문제 해결

- 그냥 값을 안넣고 있는 거였다

Malloc72P committed:
[#48] feat: 토큰 리프래시 요청을 가로채는 인터셉터 구현

Malloc72P committed:
[#48] chore: 불필요한 코드 제거

Malloc72P committed:
[#48] feat: 토큰 리프래시 기능 구현

Malloc72P committed:
[#48] feat: 컨트롤러 유틸 생성

- 인터셉터로부터 Request를 통해 전달받은 UserID를 꺼내서 반환하는 기능 추가

Malloc72P committed:
[#48] feat: 유저 인증 기능 추가

- 게임에 참여하면, 게임에 어떤 유저가 참여했는지 기록하도록 수정
- 자신이 참여한게 아닌 게임에 피치하기 요청을 하면 FORBIDDEN 에러가 발생하도록 구현함

Malloc72P committed:
[#48] feat: IntelliJ 자동 리포매팅 기능 수행

Malloc72P committed:
Merge pull request #52 from Malloc72P/BE/feature/#48/oauth_login

Be/feature/#48/oauth login 기능구현

Malloc72P committed:
chore: 주석처리된 코드 제거

Malloc72P committed:
feat : 유저 엔티티에서 지정하는 테이블 이름을 소문자로 수정

Malloc72P committed:
fix: 다시 로그인했을때 문제발생하던 부분 수정

- 기존 정보를 수정하지 않고, 아예 새로 유저정보를 만들면서 문제가 발생한 부분 수정

Malloc72P committed:
chore : IntelliJ 자동 리포매팅 수행

Malloc72P committed:
feat : 불필요한 cross origin설정 제거

ksundong commented:
커밋되면 안되는 파일이 포함된 것 같습니다.
리뷰 받고자 하는 파일만 포함해서 PR 보내주세요!

Malloc72P commented:
헉, 문제 해결한다음에 다시 요청드리겠습니다!

Malloc72P committed:
Revert "useFetch _ useState 관리 상태 1개로 정리."

Malloc72P committed:
Merge pull request #55 from Malloc72P/revert-23-fe/fixError

Revert "useFetch _ useState 관리 상태 1개로 정리."

Malloc72P commented:
해결했습니다!

Malloc72P committed:
feat: 테스트 데이터 입력 및 적용

Malloc72P committed:
fix 현재 이기고 있는 팀과 게임종료여부를 알려주도록 수정

Malloc72P committed:
fix: 테스트 데이터와 중복되는 테스트코드 수정

Malloc72P committed:
fix: 리프래시 토큰 인터셉터가 동작하지 않던 문제 해결

ksundong commented on backend/build.gradle:
```diff
@@ -0,0 +1,26 @@
+	developmentOnly 'org.springframework.boot:spring-boot-devtools'
+	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+	runtimeOnly 'mysql:mysql-connector-java'
+	implementation 'io.jsonwebtoken:jjwt:0.9.1'
```
저는 auth0의 라이브러리가 더 낫더라고요~

Malloc72P replied:
오 그러면 다음 프로젝트에서는 auth0를 써보면 좋을 것 같아요! 직접 경험해보고 어떤 차이가 있는지 배워보는것도 좋겠다는 생각이 드네요!

ksundong replied:
좋은 태도네요 👍

ksundong commented on backend/src/main/java/com/codesquad/baseball/annotation/Auth.java:
```diff
@@ -0,0 +1,11 @@
+
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Auth {
```
오호 어떤 용도의 애노테이션인지 문서화 해주면 더 좋았을 것 같아요.

Malloc72P replied:
이럴 때 주석을 활용해서 애노테이션에 대한 문서화를 해도 될까요? 주석쓰면 좋을 것 같아요!

ksundong replied:
JavaDoc은 주석이라고 보지 마시구요, 좋은 기술입니다. 적극적으로 써먹어보셔도 좋을 것 같아요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/config/AuthInterceptor.java:
```diff
@@ -0,0 +1,64 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class AuthInterceptor implements HandlerInterceptor {
```
인터셉터와 필터는 각각 어느 상황에서 사용해야 할까요?
이것도 제가 받았던 기술 면접 질문중에 하나입니다. 한 번 생각해보세요!

Malloc72P replied:
이번 기회를 이용해서 필터와 인터셉터에 대해 조금 공부해보고 정리해봤어요!
https://www.notion.so/Filter-Interceptor-AOP-37b1dd207abc4b91a8424f0eaca65ee1

필터와 인터셉터 둘 다 컨트롤러의 핸들러 메서드의 앞과 뒤에서 호출될 수 있지만, 구조상 필터가 인터셉터보다 먼저 요청을 받아서 필터링 할 수 있다는 성질을 숙지하고 써먹으면 좋을 것 같더라구요.

그런데 한가지 잘 모르겠는점이 생겼네요
토큰인증과 같은 일은 어디서 처리하는게 좋을까요?
찾아보니까 필터에서도 충분히 토큰인증관련 로직을 처리할 수 있겠더라구요.
그러면 필터랑 인터셉터 둘 다 토큰인증처리 로직을 처리할 수 있다면, 이 일을 더 잘할 수 있는 객체한테 맡기는게 좋겠는데, 이 판단을 하기 위한 조건에 어떤게 있을지 잘 모르겠어요.

우선 제 생각에는 인터셉터에서 하는편이 좋지 않을까 싶어요. 제가 이해한게 맞다면, 인터셉터도 DispatcherServlet의 ApplicationContext안에 있는 Bean이고, 그러면 여기서 예외가 발생하더라도 ControllerAdvice를 통해서 핸들링 할 수 있으니까 좀 더 편하게 구현할 수 있지 않을까 라는 생각이 들었기 때문이예요.
그러면 필터는 어떨까 싶긴 한데, 지금까지 프로젝트를 하면서 필터를 활용해본적이 없어서 잘 모르겠네요 ㅋㅋ
다음 프로젝트에서 실험적으로 필터를 사용해보면 좋겠다는 생각이 듭니다. 직접 경험해보고 어떤 불편함이 있는지, 아니면 어떤 장점이 있는지 직접 겪어봐야겠습니다

ksundong replied:
제일 중요한게 빠졌네요.
Filter는 Servlet의 스펙이고, Interceptor는 Spring의 스펙입니다.
https://goddaehee.tistory.com/154
이 글도 한 번 참고해보세요.

ksundong replied:
인증이 Spring의 도움이 필요한 경우 Interceptor일 것이고, 그게 필요없다면 Filter의 영역이겠죠?
더 나아가서, Spring Security를 사용한다면 그런 고민도 깔끔하게 해소시킬 수 있고요.
필터에서도 얼마든지 ControllerAdvice로 Handling 할 수 있기 때문에 말씀하신 이유로 선택하는건 조금 곤란할 것 같아요.
한 번 직접 필터를 사용해보세요~

ksundong commented on backend/src/main/java/com/codesquad/baseball/config/AuthInterceptor.java:
```diff
@@ -0,0 +1,64 @@
+    public static final String USER_ID_KEY = "USER_ID";
+    private static final int BEARER_TOKEN_LENGTH = 2;
+    private static final int TOKEN = 1;
+    private final JwtVerifier jwtVerifier;
```
```suggestion
    private static final int TOKEN = 1;
    
    private final JwtVerifier jwtVerifier;
```

ksundong commented on backend/src/main/java/com/codesquad/baseball/config/AuthInterceptor.java:
```diff
@@ -0,0 +1,64 @@
+        HandlerInterceptor.super.afterCompletion(request, response, handler, ex);
+    }
+
+    private void onAuthAnnotation(HttpServletRequest request) {
+        AccessTokenDTO accessTokenDTO = extractAccessTokenDtoFromRequest(request);
+        jwtVerifier.verifyAccessToken(accessTokenDTO);
+        request.setAttribute(USER_ID_KEY, accessTokenDTO.getUserId());
+    }
```
private 메서드는 사용되는 메서드 아래에 위치시켜주는 것이 가독성에 더 좋습니다.

ksundong replied:
또한 private 메서드가 너무 많이 등장한다면
별도의 클래스를 빼줘야하는 신호라고 보시면 좋을 것 같아요.

Malloc72P replied:
구현에 너무 급급해서 하나의 클래스에 너무 많은 기능을 넣어버린 것 같아요... 분리해보겠습니다!

ksundong commented on backend/src/main/java/com/codesquad/baseball/config/RefreshInterceptor.java:
```diff
@@ -0,0 +1,63 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class RefreshInterceptor implements HandlerInterceptor {
```
RefreshToken은 JWT 스펙은 아닙니다. OAuth 스펙이죠.
이것에 대해서도 공부해보시면 좋을 것 같네요.

Malloc72P replied:
로그인 기능 구현하면서 참 어려웠던게 이런 부분 같아요.
내가 하는 방법이 맞나? 하는 의문은 있는데, 찾아도 잘 안나오고 해서 어렵더라구요.

근데 이제보니 리프래시하는건 JWT쪽 이야기가 아니어서 그랬겠다는 생각이 드네요

그러면 한가지 궁금한게 있어요. 일단 제가 이해한 OAuth 로그인은 아래와 같습니다

1. 유저가 우리 서버한테 로그인 페이지를 요청
2. 우리 서버는 유저를 구글 로그인 페이지로 리다이렉션 시킴
3. 유저가 구글 로그인 페이지에서 로그인을 하고, 권한을 승인함
4. 3번에 대한 결과로 유저는 Authorization Code를 획득하고 CallBack URL로 리다이렉션됨
5. 우리 서버는 Authorization Code를 받아서 구글 인증서버에 AccessToken을 요청해서 획득함
6. 획득한 AccessToken을 가지고 유저의 프로필 정보를 요청해서 획득함
7. 여기까지해서 사용자 인증을 완료함.

저는 5번을 통해서 획득한 AccessToken은 유저의 프로필정보를 요청하는데만 쓰고 버린다고 생각했어요.
이걸 가지고 우리 서버에 대한 인증을 하면 위험하지 않을까 하는 생각때문이었어요.
그래서 따로 서버에 비밀키를 만들고, 이 키를 활용해서 JWT토큰을 만들었습니다.

제가 궁금한건, 이렇게 하는게 맞는지를 잘 모르겠어요. 
그냥 5번을 통해서 획득한 AccessToken을 써도 될까요?
아니면 제가 한대로 JWT토큰을 따로 만들어서 쓰는게 맞는걸까요?

그리고 이 경우에선 토큰 리프래시를 어떻게 하는게 좋을지를 잘 모르겠어요.
구글한테서 얻은 액세스토큰과 리프래시 토큰을 활용해서 하는건지, 아니면 제가 구현한 JWT 토큰 리프래시 방법도 괜찮은건지, 이걸 잘 모르겠네요. 아니면 이 질문에 대한 해답을 찾으려면 어떤걸 읽어보면 좋을까요?

ksundong replied:
잘 생각해봅시다. 구글에서 준 AccessToken은 어떤 AcessToken일까요?
우리가 발급해 준 JWT Token은 어떤 AccessToken일까요?

---

토큰 리프레시는 OAuth 스펙이라고 말씀드린 부분인데, 잘 생각해보세요.
토큰 리프레시 하는 방법은 찾아보시면 꽤 나올거에요. 다만 저는 리프레시 토큰을 위해 JWT를 쓰는건 바람직하지 않다고 생각해서 회사에 구현한 리프레시 로직은 UUID + Cache Server로 구현되어 있습니다.
질문에 대한 답은 여러개이기 때문에 많은 생각 혹은 경험을 해보시길 추천드려요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/config/RefreshInterceptor.java:
```diff
@@ -0,0 +1,63 @@
+        request.setAttribute(USER_ID_KEY, refreshTokenDTO.getUserId());
+    }
+
+    private RefreshTokenDTO verifyRefreshToken(HttpServletRequest request) {
+        String jwtToken = TokenUtil.extractTokenFromHeader(request);
+        String userId = jwtVerifier.extractUserIdFromJwt(jwtToken);
+        return new RefreshTokenDTO.Builder()
+                .userId(userId)
+                .refreshToken(jwtToken)
+                .build();
+    }
```
RefreshToken의 본질은 무엇일까요?

Malloc72P replied:
보안을 강화하면서 사용자 경험을 최대한 해치지 않도록 도와주는 장치라고 생각해요

액세스토큰만 있으면 사용자 인증을 할 수 있으니까 노출되었을때 많이 위험할 것 같아요.
그래서 액세스토큰의 수명을 짧게 설정하는 것 같더라구요.

하지만 이렇게 하면 액세스토큰이 만료될때마다 사용자가 번거로운 로그인 작업을 반복해야 하니까 사용자 경험적인 측면에서는 좋지 않을 것 같아요.
그대신 리프래시 토큰을 적용하면 되는 것 같습니다. 액세스토큰보다 상대적으로 수명이 긴 리프래시 토큰을 발급하고, 이걸로 액세스토큰의 수명을 연장시킨다면 사용자가 매번 로그인작업을 반복하지 않아도 되니까 좋을 것 같아요.

그런데 생각해보니까 제 프로그램은 액세스토큰의 만료여부와 관계없이 리프래시토큰으로 리프래시 요청만하면 새로운 토큰을 발급해주네요? 음... 액세스토큰이 만료되었는지도 검사해줘야 할 것 같아요.

근데 이것도 맞는 설계인지 잘 모르겠어요... 위에서 언급하신 내용처럼, 토큰 리프래시는 JWT의 스펙이 아니니까, 이런 방식의 구현은 JWT토큰의 설계의도와 맞지 않는다는 생각이 드네요...

ksundong replied:
해당 리프레시 토큰은 그럼 노출이 안될까요?
리프레시 토큰을 가지고 액세스 토큰의 갱신 요청만 해도 말씀하신 위험이 발생할텐데요!
GitHub같은 경우에는 Opaque Token을 사용하고, Refresh Token을 발급하지 않습니다.
왤까요? 저도 생각해보지 않은 문제인데, 한 번 생각해보시면 좋을 것 같네요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/config/WebMvcConfig.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
+        registry.addInterceptor(new AuthInterceptor(jwtVerifier))
+                .addPathPatterns("/**")
+                .excludePathPatterns("/users/**");
+
+        registry.addInterceptor(new RefreshInterceptor(jwtVerifier))
+                .addPathPatterns("/users/**");
+    }
```
https://eastglow.github.io/back-end/2019/08/01/Spring-Interceptor-사용-시-의존성-주입이-안되는-경우.html

interceptor도 bean으로 관리해보면 어떨까요?

Malloc72P replied:
와... 이게 되네요...... 꿀팁 정말 감사합니다!

Malloc72P replied:
그런데 이렇게하면 인터셉터에서는 생성자주입을 못쓰게 되는건가요?

그러면 WebMvcConfigurer에서 직접 인터셉터의 생성자에 서비스를 주입하는 방법이 더 좋지 않을까 하는 생각이 들어요

Malloc72P replied:
음 근데 좀 더 고민해보니까 생성자 주입으로 하면 WebMvcConfig에서 매번 번거롭게 주입해줘야 하네요...
그냥 인터셉터도 빈으로 관리해서 Resource애너테이션으로 의존성을 주입하는게 더 좋을 것 같아요!

ksundong replied:
생성자 주입은 됩니다.
아직 스프링을 잘 이해하지 못하신게 아닐까 하는 생각도 드는군요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/ExceptionController.java:
```diff
@@ -0,0 +1,61 @@
+@RestControllerAdvice
+public class ExceptionController {
+
+    @ResponseStatus(HttpStatus.CONFLICT)
```
좋은 에러코드입니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/ExceptionController.java:
```diff
@@ -0,0 +1,61 @@
+    @ResponseStatus(HttpStatus.CONFLICT)
+    @ExceptionHandler(GameAlreadyOccupiedException.class)
+    public ErrorDTO handleGameAlreadyOccupiedException(GameAlreadyOccupiedException exception) {
+        return new ErrorDTO(HttpStatus.CONFLICT, exception.getMessage());
```
status를 리턴할 필요가 있나요?

Malloc72P replied:
아 이부분은 사이드디쉬 프로젝트를 진행할때 모바일 팀원분이 바디로 상태코드도 주면 좋겠다고 하신게 생각이 나서 반영했던 부분이예요. 제가 모바일이랑 리엑트쪽 상황을 잘 몰라서 일단 넣어주도록 했습니다.

ksundong replied:
원래라면 필요 없어야 합니다. ㅎㅎ

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/ExceptionController.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler({PlayerNotFoundException.class, TeamNotFoundException.class, GameNotFoundException.class, UserNotFoundException.class})
```
그냥 NotFoundException으로 핸들링했으면 안될까요?

Malloc72P replied:
오... 저거 설마 `{A.class, B.class, C.class}`이렇게 안해도 되는거예요?
몰랐어요... 꿀팁 감사합니다! 한번 해보고 다시 코멘드 달겠습니다!

ksundong replied:
기다리고 있겠습니다. ㅎㅎ

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/ExceptionController.java:
```diff
@@ -0,0 +1,61 @@
+
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler({PlayerNotFoundException.class, TeamNotFoundException.class, GameNotFoundException.class, UserNotFoundException.class})
+    public ErrorDTO handlePlayerNotFoundException(NotFoundException exception) {
```
메서드 명 확인해주세요!

Malloc72P replied:
으악 실수했네요!

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/ExceptionController.java:
```diff
@@ -0,0 +1,61 @@
+
+    @ResponseStatus(HttpStatus.METHOD_NOT_ALLOWED)
+    @ExceptionHandler(HttpRequestMethodNotSupportedException.class)
+    public ErrorDTO handleInternalServerError(HttpRequestMethodNotSupportedException exception) {
```
여기도 메서드명 확인해주세요!

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/ExceptionController.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ExceptionHandler({NoJwtTokenException.class, InvalidJwtTokenException.class})
+    public ErrorDTO handleNoJwtTokenException(RuntimeException exception) {
```
둘을 묶는 다른 Exception을 선언해볼까요?

Malloc72P replied:
NotFoundException처럼 상위 클래스를 하나 만들어주면 그룹화해서 관리할 수 있겠네요!

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/ExceptionController.java:
```diff
@@ -0,0 +1,61 @@
+        return new ErrorDTO(HttpStatus.CONFLICT, exception.getMessage());
+    }
+
+    @ResponseStatus(HttpStatus.FORBIDDEN)
+    @ExceptionHandler(NotYourGameException.class)
+    public ErrorDTO handleNotYourGameException(NotYourGameException exception) {
+        return new ErrorDTO(HttpStatus.FORBIDDEN, exception.getMessage());
+    }
```
👍

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/GameController.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    @Auth
+    @PatchMapping("/{gameId}")
```
왜 Patch를 써주셨는지 궁금해요.
이유가 있을 것 같거든요.

Malloc72P replied:
아, 이 부분은 Put이랑 Patch중에서 고민하다가 Patch를 선택했어요.

제가 이해한 Patch와 Put메서드는 아래와 같습니다
- Put은 타겟이 되는 리소스가 없으면 보낸 데이터대로 생성하고, 있으면 교체하는식의 수정요청을 할때 사용
- Patch는 리소스에 대한 부분적인 수정을 요청할때 사용

이렇게 생각해봤을 때, 해당 컨트롤러 메서드는 게임의 선점상태값만 부분수정하는 기능이니까 Patch가 맞겠다는 생각을 했습니다. 또, 대상 게임이 없을때 게임을 생성한다던가 하는 기능도 없으니까 더더욱 Put을 쓰면 안되지 않을까? 하는 생각이 들더라구요.

ksundong replied:
네 Patch는 Partial Update를 의미하는게 맞습니다.
잘 이해하고 계시네요. 😊

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/GameController.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Auth
+    @PatchMapping("/{gameId}")
+    public void joinIn(@PathVariable("gameId") int gameId, HttpServletRequest request) {
```
HttpServletRequest를 받지 않는 방식으로 개발하려면 어떻게 해야할까요?

Malloc72P replied:
아, 이부분도 엄청 고민하던 부분 중 하나예요.
저도 `HttpServletRequest`말고 다른게 없을까? 해서 이것저것 시도도 해보고 구글링도 해봤는데...
결국 이것말고는 못찾았어요...

여기서 관건은 인터셉터에서 컨트롤러 메서드로 값을 넘겨주는 중간 개체가 있거나 하면 좋겠는데, HttpServletRequest말고는 잘 모르겠더라구요....

근데 지금 다시보니까 인터셉터에서 handler를 받잖아요? 이거가지고 어떻게 해볼 수 없을까? 라는 생각이 드네요.
한번 도전해보고 다시 코멘트 남기겠습니다!

Malloc72P replied:
좀 더 찾아봤는데 도저히 모르겠어요......
인터셉터에서 컨트롤러의 핸들러메서드로 토큰에서 꺼낸 UserID를 넘겨주고 싶은데, handler는 값을 담아서 갈 수 있는 객체가 아닌 것 같더라구요...
아니면 이런 방식의 구조는 피해야하는걸까요?

인터셉터에서 핸들러로 값을 넘기는걸 피해야 한다면, 핸들러 메서드에서 request의 헤더에 있는 토큰을 꺼내서 해독하고 userId를 꺼내는 일을 다시 수행해야할 것 같은데, 인터셉터에서 이미 수행한 동작을 또 하는것도 별로 좋은 방법은 아닌 것 같아요.

그러면 생각나는게 request랑 response말고는 없는데, 혹시 힌트 주실 수 없을까요?

ksundong replied:
https://github.com/ksundong/simple-spring-param-resolver
여길 참고해보세요.

Malloc72P replied:
와 제가 지금까지 아무생각없이 쓰고있던 `@PathVariable`이 디온이 알려주신 ArgumentResolver를 통해서 이루어지는거였군요... 배운 내용을 노션에 정리해보았습니다
https://www.notion.so/HandlerMethodArgumentResolver-5ebd3295d13349d2a0548d96f73a8054

baseball프로젝트에도 적용해봤어요! 
디온이 컨트롤러의 핸들러메서드에서 `HttpServletRequest`를 안받고 구현하라고  말씀하신 이유를 조금은 알 것 같아요.
어떻게 보면 `PathVariable`도 이거랑 같은 맥락이 아닐까 싶어요. url에서 pathVariable을 얻고자 한다면, 직접 HttpServletRequest를 받아서 처리하는 방법도 있겠죠. 하지만 이런식으로 request에서 무언가를 꺼내는 일이 많아진다면, 컨트롤러에서 이걸 꺼내는 메서드를 호출하는것도 불필요한 작업이 되겠다는 생각이 드네요. 그냥 ArgumentResolver를 쓰면 해결되는 문제니까요.

알려주셔서 정말 감사합니다!

ksundong replied:
PathVariable의 원리는 저도 잘 모르겠네요. 비슷한 로직이겠죠 ㅎㅎ
저도 이론만 알고 실제로 구현하는데 한 시간 정도 걸렸는데, 도움이 되었으면 좋겠네요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/GameController.java:
```diff
@@ -0,0 +1,47 @@
+    @Auth
+    @PatchMapping("/{gameId}")
+    public void joinIn(@PathVariable("gameId") int gameId, HttpServletRequest request) {
+        String userId = ControllerUtil.extractUserIdFromRequest(request);
```
Util의 사용은 꼭 필요한 경우가 아니라면 지양해주세요.

Malloc72P replied:
앗 이부분은 중복코드를 줄이려는 목적으로 사용한건데요, 그러면 이런 경우에는 어떤 방법을 사용하는게 좋을까요?

유틸 사용을 피한다면, 서비스로 만들어보는것도 좋을 것 같아요!

저 근데 유틸사용을 지양해야 하는 이유가 있을까요? 요 부분도 스테틱 사용을 최대한 피해야 하는거랑 비슷한 맥락일까요?

ksundong replied:
네 맞습니다. 객체지향적인 사고를 유도하기 위함입니다.
적절한 사용은 말리지 않습니다만, 먼저 Util을 생각하지말고, Service 내지는 Component로 객체가 그 일을 담당하게 끔 생각해봐주세요.
그리고, 이제는 필요없어진 코드리라 생각됩니다. 😊

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/GameController.java:
```diff
@@ -0,0 +1,47 @@
+        return gameService.gameDetail(gameId);
+    }
+
+    @Auth
```
모든 메서드에 붙이는 것은 비효율적이지 않을까요?

Malloc72P replied:
오... 이 부분도 코드 중복이 되겠네요...
이 부분은 WebMvcConfigure에서 설정을 좀 더 잘해주면 `@Auth`를 안붙이고도 해결할 수 있을 것 같아요.
한번 도전해보겠습니다!

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/UserController.java:
```diff
@@ -0,0 +1,56 @@
+@RequestMapping("/users")
+public class UserController {
+
+    private final GoogleApiRequester googleApiRequester;
+    private final UserService userService;
+    private final GoogleUrlService googleUrlService;
+    private final JwtBuilder jwtBuilder;
```
객체로 많이 분리가 되었군요!

Malloc72P replied:
저번 사이드디쉬 프로젝트에서 가져와서 조금 고쳐봤습니다!  이정도면 잘 분리한거 맞죠? 😄

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/UserController.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @GetMapping("/login")
+    public void login(HttpServletResponse response) throws IOException {
+        response.sendRedirect(googleUrlService.loginPageUrl());
+    }
```
300번대 응답을 직접 주시면 어땠을까요?

Malloc72P replied:
아하 [여기 내용처럼](https://developer.mozilla.org/ko/docs/Web/HTTP/Status/304) 304로 응답을 주는편이 더 좋겠네요!
`캐시된 자원으로의 암묵적인 리다이렉션이다`라는걸 상태코드를 가지고 표현할 수 있어서 이렇게 해야겠네요!
알려주셔서 감사합니다!

ksundong commented on backend/src/main/java/com/codesquad/baseball/controller/UserController.java:
```diff
@@ -0,0 +1,56 @@
+    @GetMapping("/callback")
+    public JwtTokenDTO oauthCallBack(AuthorizationInfo authorizationInfo) {
+        ReceiveAccessTokenDTO receiveAccessTokenDTO = googleApiRequester.requestAccessToken(authorizationInfo.getCode());
+        UserInfoDTO userInfoDTO = googleApiRequester.requestUserInfo(receiveAccessTokenDTO.getAccess_token());
```
언더바가 등장했네요..?

Malloc72P replied:
앗, 이거 저번 사이드디쉬때 달아주신 리뷰인데 깜빡하고 못고쳤네요... 정말 죄송합니다 ㅠㅠ

약간 변명을 하자면 구글에서는 스네이크 케이스로 응답을 주더라구요. 여기에 맞추느라고 스네이크 케이스로 했던건데, 보니까 백엔드 동료인 코코가 이런것도 설정을 잘 하면 Jackson에서 스네이크케이스로 온 json을 카멜케이스로 잘 바꿔준다고 했던게 기억나네요. 카멜케이스로 바꿔도 잘 돌아갈 수 있도록 고쳐봐야겠네요

Malloc72P replied:
```java
@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
public class ReceiveAccessTokenDTO {

    private String accessToken;
    private int expiresIn;
    private String scope;
    private String tokenType;
    private String idToken;
}
```
JsonNaming애노테이션으로 해결했습니다!

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+    private static final int LAST_EXTENDED_INNING = 12;
+    private static final boolean END_OF_THE_GAME = true;
+
+    @Id
+    private Integer id;
+    private String gameTitle;
+    private boolean isTop;
+    private int currentStrikeCount;
+    private int currentOutCount;
+    private int currentBallCount;
+    private String gameOccupier;
+    private Set<TeamParticipatingInGame> teams = new HashSet<>();
+    private int firstBase;
+    private int secondBase;
+    private int thirdBase;
+    @MappedCollection(idColumn = "game", keyColumn = "inning_number")
+    private List<Inning> innings = new ArrayList<>();
+    @Transient
+    private boolean isInitialized = false;
```
객체가 너무 큽니다.
적절한 단위로 분리하는게 좋을 것 같네요.

Malloc72P replied:
맞아요 이거 코드 짜면서도 너무 커서 분리하고 싶었는데 방법이 떠오르지 않아서 그냥 했던 기억이 나요.

조금 시간이 지나서 보니까, 게임 객체를 이렇게 분리하면 어떨까 하는 생각이 드네요

```
Game
  |______ TeamData(팀 정보관리)
  |______ Diamond(1,2,3루랑 마운드, 타석 상태정보 관리)
  |______ DashBoard(이닝정보 및 점수관리)
```
하 근데 이거 또 걱정이 되는게 SpringDataJDBC를 쓰다보니까 설정하나 잘못건들면 막 안되고 그러니까 객체분리하는게 엄청 겁나네요 ㅋㅋㅋㅋ
그래도 한번 해보고 싶네요. 도전해보겠습니다

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+        }
+    }
+
+    public List<Map<TeamType, Integer>> showScoreBoard() {
```
Map이 아니라 객체의 List였다면 어떨까요?

Malloc72P replied:
아 이 부분을 Map으로 한 이유는, 팀이 홈인지 원정팀인지 여부를 표현하기 위해서였어요.

그랬긴한데, 이거 그냥 제가 야구 규칙을 잘 몰라서 이렇게 한 거였더라구요 😄 

이거 점수판에서 위가 away이고 아래가 home인 것 같더라구요. 그러면 그냥 List로 표현해서 0번째를 away, 1번째를 home으로 하는편이  좀 더 야구규칙에 맞겠다는 생각이 드네요 😄 

이게 다 야구를 게임으로만 하고 직접 본 적이 없어서 그런 것 같습니다 ㅎㅎㅎㅎ 너무 어릴때 했던거라 기억이 잘 안나네요

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+    }
+
+    public TeamType winner() {
+        int homeTeamScore = homeTeamScore();
+        int awayTeamScore = awayTeamScore();
+        if (homeTeamScore > awayTeamScore) {
+            return TeamType.HOME;
+        } else if (homeTeamScore < awayTeamScore) {
+            return TeamType.AWAY;
+        } else {
+            return TeamType.NONE;
+        }
```
승자를 결정하는 책임을 다른 객체에 넘겨주는게 어떨까 싶네요.

Malloc72P replied:
오, 이것도 객체분리할때 적용해보면 좋겠네요.

그러면 경기판정하는 로직을 분리해볼 수 있겠어요!

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+
+    public PitchResult pitch(PlayType playType) {
+        PitchResult pitchResult = new PitchResult(playType);
+        switch (playType) {
+            case STRIKE:
+                pitchResult = onStrike();
+                break;
+            case BALL: {
+                pitchResult = onBall();
+                break;
+            }
+            case HITS: {
+                pitchResult = onHits();
+                break;
+            }
+            case HOMERUN: {
+                pitchResult = onHomeRun();
+                break;
+            }
+        }
```
switchCase를 쓰지않고, 로직을 전개시켜보세요.
PlayType이 해당 기능을 가지게 한다던가하는 식으로 바꿀 수 있어보입니다.

Malloc72P replied:
오 그러면 `playType.play()` 이런식으로 하면 스위치 케이스문이 한줄짜리 코드가 될 수 있겠군요!

근데 저기있는 onStrike, onBall같은 메서드들은 Game의 상태값을 바꾸는 메서드인데, 저게 PlayType으로 넘어가면 PlayType에서 외부의 객체인 Game의 상태값을 바꾸는게 되버려서 위험해질 것 같아요. 아니면 다른 방법이 있을까요?

야구 로직처리하는 부분이 많은 아쉬움이 남네요... 처음에는 대충 이렇게 하면 되지 않을까 해서 했던건데, 전체설계를 하지 않고 무작정 덤빈 대가를 톡톡히 치른 것 같네요...

ksundong replied:
상태값을 바꾸지 않고, 상태값을 새로 지정해서 리턴하는 방식도 있겠죠~

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+    }
+
+    private void savePitchResult(PitchResult pitchResult) {
+        History history = currentInning().createHistory(pitchResult.getPlayType(), currentStrikeCount, currentBallCount,
+                defendingTeam().getCurrentPitcher(), attackingTeam().getCurrentHitter(), pitchResult.numberOfRunners());
```
한 번 다시 생각해보는게 좋을 것 같아요.

Malloc72P replied:
속성값대신 객체를 넘기는편이 더 좋았을까요?
메서드를 만들때, 어떨때는 속성을 넘기고, 어떨때는 객체를 넘겨야 할지가 잘 감이 안잡히네요.
좀 더 고민해보겠습니다!
혹시 좋은 팁이 있다면 알려주세요!

ksundong replied:
History Builder를 써서 history를 만든다음, history를 넣어줘도 좋을 것 같네요.

ksundong replied:
그냥 파라미터가 3개 이상이다 싶으면, 객체로 묶거나 빌더를 쓰는 편이에요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+    private PitchResult onStrike() {
+        PitchResult pitchResult = new PitchResult(PlayType.STRIKE);
+        increaseStrikeCount();
+        PlayType strikeResult = judgeThreeStrike();
+        if (strikeResult == PlayType.STRIKE_OUT) {
+            pitchResult.changePlayTypeToStrikeOut();
+        }
```
이부분의 로직전개가 조금 아쉽네요.

Malloc72P replied:
처음부터 야구게임 로직을 어떻게 구현할지 잘 설계하고 들어갔어야 하는데, 그렇게 안하고 무작정 덤빈 대가를 치르고 있는 것 같네요 ㅠㅠ
맞아요 저게 요구사항 분석을 완벽하게 안하고 짜다보니까, 나중으로 가면 갈수록 코드 수정이 많이 발생하더라구요.
그러면서 이해하기도 어렵고, 복잡하고 난해한 코드가 되어버린 것 같네요... 반성하겠습니다

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/PlayType.java:
```diff
@@ -0,0 +1,11 @@
+package com.codesquad.baseball.domain.game;
+
+public enum PlayType {
+    HOMERUN,
+    STRIKE,
+    HITS,
+    BALL,
+    STRIKE_OUT,
+    FOUR_BALL,
+    AT_BAT
```
너무 종류가 다양하니 성격에 따라서 나눠주는게 좋아보입니다.

Malloc72P replied:
별도로 OutType같은걸 만들면 어땠을까 하는 생각이 드네요. 확실히 PlayType이라고 보기엔 어려운 값들이 많이 있네요

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+        return pitchResult;
+    }
+
+    private PitchResult judgePitchResult(PitchResult pitchResult) {
+        switch (pitchResult.getPlayType()) {
+            case HOMERUN:
+            case HITS:
+            case FOUR_BALL:
+                changeHitterOfAttackingTeam();
+                resetStrikeAndBall();
+                break;
+            case STRIKE_OUT:
+                changeHitterOfAttackingTeam();
+                resetStrikeAndBall();
+                increaseOutCount();
+                if (isThreeOut()) {
+                    boolean isEndOfTheGame = proceedToNextStage();
+                    if (isEndOfTheGame) {
+                        pitchResult.changeGameStateToGameOver();
+                    }
+                }
+                break;
+        }
+        pitchResult.getBackHomeRunners().forEach(i -> currentInning().addScore(attackingTeam()));
+        return pitchResult;
```
depth수는 곧 복잡도입니다.
지금은 몇 depth죠?
이렇게 많은 조건문이 들어간다면, 다른 개발자는 어떻게 이해할 수 있을까요?

ksundong replied:
이 코드의 복잡도를 파악하는 아주 간단한 방법이 있습니다.
이 코드의 동작을 테스트 코드를 작성한다 가정하고 목록을 작성해보세요.

Malloc72P replied:
에고고... 너무 배려없는 코드였군요... 조심하도록 하겠습니다

너무 많이 깊어진체로 코드를 작성해놓으니까 제가 짰는데도 보기가 어렵네요
이런 방식의 코드작성은 최대한 피하도록 하겠습니다

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+        return PlayType.STRIKE;
+    }
+
+    private PitchResult judge4Ball() {
```
왜 숫자를 써주신건지 잘 모르겠어요.

Malloc72P replied:
볼넷상황에 대한 판정을 하는 메서드라서, 볼넷을 표현하려고 4를 작성했어요

ksundong replied:
이게 자바나 프로그래밍 진영에서 4는 for를 의미하는 경우가 많아서 불필요하게 숫자를 쓰기보다는 Four라고 써주는게 더 좋아보여요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+            int backHomeRunner = pushAllRunners();
+            pitchResult.changePlayTypeToFourBall();
+            pitchResult.addRunner(backHomeRunner);
+            return pitchResult;
+        }
+        return pitchResult;
```
```suggestion
        }
        return pitchResult;
```

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+        PitchResult pitchResult = new PitchResult(PlayType.BALL);
+        if (is4Ball()) {
+            int backHomeRunner = pushAllRunners();
+            pitchResult.changePlayTypeToFourBall();
+            pitchResult.addRunner(backHomeRunner);
```
되도록이면 상태를 변경하기 보다는 새로운 정보를 리턴하는 형태를 생각해봐주세요.
그러면 좀 더 간단한 코드가 작성됩니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+        return backHomePlayer;
+    }
+
+    private List<Integer> recallAllRunners() {
+        List<Integer> backHomeRunners = new ArrayList<>();
+        if (hasThirdBaseRunner()) {
+            backHomeRunners.add(thirdBase);
+            thirdBase = NO_PLAYER;
+        }
+        if (hasSecondBaseRunner()) {
+            backHomeRunners.add(secondBase);
+            secondBase = NO_PLAYER;
+        }
+        if (hasFirstBaseRunner()) {
+            backHomeRunners.add(firstBase);
+            firstBase = NO_PLAYER;
+        }
+        backHomeRunners.add(attackingTeam().getCurrentHitter());
+        return backHomeRunners;
```
모든 주자가 돌아와야 하는 경우는 결국 새로운 상황에서 진행되는 것이라고 생각해보면 어떨까요?

Malloc72P replied:
그것도 좋은 방법일 것 같아요!

일단 위의 코드는 클라이언트한테 홈에 들어온 주자들의 타순과 들어온 순서까지 주고 싶어서 저렇게 구현하게 되었습니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+        return thirdBase != NO_PLAYER;
+    }
+
+    public int firstBaseRunner() {
+        return firstBase;
+    }
+
+    public int secondBaseRunner() {
+        return secondBase;
+    }
+
+    public int thirdBaseRunner() {
+        return thirdBase;
+    }
```
int가 의미하는게 너무 추상적이어서, 별도의 객체가 있었으면 하네요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+        return defendingTeam().getCurrentPitcher();
+    }
+
+    private boolean proceedToNextStage() {
+        resetField();
+        if (isTop) {
+            isTop = false;
+        } else {
+            if (judgeGameOver()) {
+                return END_OF_THE_GAME;
+            }
+            proceedToNextInning();
+        }
+        return !END_OF_THE_GAME;
+    }
```
간소화해봅시다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+    }
+
+    private boolean judgeGameOver() {
+        if (isLastInning() && !isTieState()) {
+            return END_OF_THE_GAME;
+        } else if (isExtendedInning() && !isTieState()) {
+            return END_OF_THE_GAME;
+        } else if (isLastExtendedInning()) {
+            return END_OF_THE_GAME;
+        }
```
각각을 변수로 추출하고 OR 조건으로 만들면 어떨까요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+        return currentInningNumber() == LAST_INNING;
+    }
+
+    private boolean isExtendedInning() {
+        int inningNumber = currentInningNumber();
+        return LAST_INNING < inningNumber && inningNumber <= LAST_EXTENDED_INNING;
+    }
+
+    private boolean isLastExtendedInning() {
+        return currentInningNumber() == LAST_EXTENDED_INNING;
+    }
```
연장 경기까지 설계한건 너무 어려움이 따랐을 것 같습니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/Game.java:
```diff
@@ -0,0 +1,610 @@
+            return new Game(this);
+        }
+    }
+}
```
이 클래스의 라인수는 610개고 메서드의 수는 엄청 많습니다.
이런 클래스는 책임도 너무많고 변경이 발생할 때, 헷갈릴 여지가 아주 많습니다.
꼭 개선해주세요. 다음 프로젝트에서 이런 클래스는 작성하지 않는것을 강력하게 권장하겠습니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/inning/Inning.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Id
+    private Integer id;
+    private int game;
```
game은 무엇인데 int 타입이죠?

Malloc72P replied:
game에 대한 id입니다

DataJdbc에서 OneToMany관계로 할때, 스키마로 상위 어그리게이트에 대한 fk의 이름을 테이블로 쓰더라구요.
바꾸면 game말고 다른걸로 쓸 수 있긴 한데 그러지 않고 그냥 쓴것으로 보입니다
다음부터는 MappedCollection을 사용해서 좀 더 명확한 이름을 주도록 하겠습니다. game보다는 gameId가 더 좋겠네요

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/inning/Inning.java:
```diff
@@ -0,0 +1,88 @@
+    protected Inning() {
+    }
+
+    public Inning(int homeTeamScore, int awayTeamScore) {
+        this.homeTeamScore = homeTeamScore;
+        this.awayTeamScore = awayTeamScore;
+    }
```
이렇게 생성하면 inningNumber와 game은 0으로 초기화 될겁니다.
원하는게 그게 맞나요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/inning/Inning.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    public void addScore(TeamParticipatingInGame attackingTeam) {
+        TeamType teamType = attackingTeam.getTeamType();
```
그냥 attackingTeam이 점수를 가지고 있으면, 그냥 그 팀의 점수를 올리면 될텐데요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/participant/PitcherPosition.java:
```diff
@@ -0,0 +1,5 @@
+package com.codesquad.baseball.domain.game.participant;
+
+public enum PitcherPosition {
+    SP, RP, CP, NONE
```
NONE은 왜 필요한가요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/participant/PlayerParticipatingInGame.java:
```diff
@@ -0,0 +1,95 @@
+    private int outCount;
+    private PitcherPosition pitcherPosition;
+
+    public PlayerParticipatingInGame() {
+    }
```
비어있는 생성자를 선언해주신 이유가 있나요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/participant/PlayerParticipatingInGame.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Id
+    private Integer id;
+    private int team_participating_in_game;
```
자바 컨벤션을 지켜주세요...

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/participant/PlayerParticipatingInGame.java:
```diff
@@ -0,0 +1,95 @@
+        this.pitcherPosition = pitcherPosition;
+    }
+
+    public float avg() {
```
줄여쓰기는 지양해주세요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/participant/PlayerParticipatingInGame.java:
```diff
@@ -0,0 +1,95 @@
+        if (plateAppearances == 0) {
+            return 0.0f;
+        }
+        return (float) hitCount / (float) plateAppearances;
```
둘 중 한 쪽만 형변환 해줘도 될텐데요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/participant/PlayerParticipatingInGame.java:
```diff
@@ -0,0 +1,95 @@
+
+import org.springframework.data.annotation.Id;
+
+public class PlayerParticipatingInGame {
```
어떤 용도의 클래스인지 잘 짐작이 안되네요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/participant/TeamParticipatingInGame.java:
```diff
@@ -0,0 +1,131 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class TeamParticipatingInGame {
```
이런것까지 도메인이어야 하나? 싶어요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/game/pitch/PitchResult.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    public void addRunner(int runner) {
+        if (runner == Game.NO_PLAYER) {
+            return;
+        }
```
잘 이해가 되지 않는 로직이네요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/team/Team.java:
```diff
@@ -0,0 +1,64 @@
+    @Id
+    private Integer id;
+    private String teamName;
+    private Set<Player> players = new HashSet<>();
```
Set을 사용해주신 이유가있나요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/team/TeamType.java:
```diff
@@ -0,0 +1,7 @@
+public enum TeamType {
+    HOME,
+    AWAY,
+    NONE,
```
NONE팀은 무엇인가요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/user/User.java:
```diff
@@ -0,0 +1,88 @@
+        locale = userInfoDTO.getLocale();
+        this.accessToken = jwtTokenDTO.getAccessToken();
+        this.refreshToken = jwtTokenDTO.getRefreshToken();
+    }
```
Domain이 DTO를 아는 방식의 설계는 지양해주세요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/user/User.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Id
+    private Long id;
+    private String userId;
+    private String email;
+    private String verifiedEmail;
+    private String name;
+    private String familyName;
+    private String givenName;
+    private String picture;
+    private String locale;
+    private String accessToken;
+    private String refreshToken;
```
필드를 적절히 책임별로 객체를 나눠보는건 어떨까요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/domain/user/User.java:
```diff
@@ -0,0 +1,88 @@
+        givenName = userInfoDTO.getGiven_name();
+        picture = userInfoDTO.getPicture();
+        locale = userInfoDTO.getLocale();
+    }
```
Update용 DTO를 별도로 선언해서 관리하는 것을 추천드릴게요.

ksundong replied:
흔히 Converter라고 불리는 형태의 구현입니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/dto/etc/ErrorDTO.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.http.HttpStatus;
+
+public class ErrorDTO {
+    HttpStatus status;
```
접근제어자가 누락되었네요.

ksundong replied:
근데 HttpStatus는 왜 필드로 가지고 있나요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/dto/game/GameDTO.java:
```diff
@@ -0,0 +1,91 @@
+        return awayTeam;
+    }
+
+    public static class Builder {
```
빌더정도는 롬복을 써줘도 괜찮다고 생각해요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/dto/game/GameScoreDTO.java:
```diff
@@ -0,0 +1,91 @@
+
+    public static GameScoreDTO from(Game game) {
+        List<InningScoreDTO> inningScoreDTOS = game.getInnings().stream()
+                .map(inning -> new InningScoreDTO.Builder()
+                        .inning(inning.getInningNumber())
+                        .homeTeamScore(inning.getHomeTeamScore())
+                        .awayTeamScore(inning.getAwayTeamScore())
+                        .build())
```
빌더를 쓰지말고, 정적 팩토리 메서드를 사용해보시면 더 깔끔해질겁니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/dto/oauth/JwtTokenDTO.java:
```diff
@@ -0,0 +1,82 @@
+public class JwtTokenDTO {
+
+    private String userId;
+    private String accessToken;
+    private String refreshToken;
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
+    private Date accessTokenExpirationDate;
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
+    private Date refreshTokenExpirationDate;
```
각각을 객체로 만들어주면 어떨까요?

ksundong replied:
그리고 Java8의 날짜 시간 API를 사용해주세요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/dto/oauth/JwtTokenResponse.java:
```diff
@@ -0,0 +1,16 @@
+public class JwtTokenResponse {
+    private String token;
+
+    public JwtTokenResponse() {
+    }
```
이렇게 하면 token이 null인 상황이 발생할 수 있을 것 같은데요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/dto/oauth/ReceiveAccessTokenDTO.java:
```diff
@@ -0,0 +1,44 @@
+    private int expires_in;
+    private String scope;
+    private String token_type;
+    private String id_token;
```
자바의 네이밍 컨벤션을 준수하는 형태로 변경해주세요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/dto/oauth/RequestAccessTokenDTO.java:
```diff
@@ -0,0 +1,43 @@
+    private String client_secret;
+    private String code;
+    private String grant_type;
+    private String redirect_uri;
```
여기도 마찬가지로 네이밍 컨벤션 지켜주세요!

ksundong commented on backend/src/main/java/com/codesquad/baseball/dto/oauth/ServerSecretDTO.java:
```diff
@@ -0,0 +1,22 @@
+    private String clientSecret;
+    private String serverSecret;
+
+    public ServerSecretDTO() {
+    }
```
기본생성자만 세팅하면 나머지 필드는 어떻게 채워지나요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/dto/oauth/UserInfoDTO.java:
```diff
@@ -0,0 +1,61 @@
+    private String given_name;
+    private String picture;
+    private String locale;
+
```
여기도 네이밍 컨벤션 지켜주세요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/exceptions/game/GameAlreadyOccupiedException.java:
```diff
@@ -0,0 +1,11 @@
+
+import com.codesquad.baseball.exceptions.notfound.NotFoundException;
+
+public class GameAlreadyOccupiedException extends NotFoundException {
```
NotFoundException하고 큰 연관이 없는 것 같은데요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/exceptions/game/GameIsNotStartedException.java:
```diff
@@ -0,0 +1,11 @@
+
+import com.codesquad.baseball.exceptions.notfound.NotFoundException;
+
+public class GameIsNotStartedException extends NotFoundException {
```
여기도 마찬가지로 NotFoundException하고는 큰 연관이 없어보입니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/exceptions/notfound/NotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package com.codesquad.baseball.exceptions.notfound;
+
+public class NotFoundException extends RuntimeException {
```
예외 나누는걸 주저하지 마세요.
NotFoundException이 책임져야하는 범위만 책임져야 하는게 더 직관적이지 않을까요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/exceptions/oauth/ExtractUserIdFromTokenFailedException.java:
```diff
@@ -0,0 +1,9 @@
+package com.codesquad.baseball.exceptions.oauth;
+
+public class ExtractUserIdFromTokenFailedException extends RuntimeException {
+    public static final String EXTRACTION_FAILED = "인터셉터로부터 UserID를 전달받지 못했습니다";
```
메시지에 굳이 인터셉터라는 표현이 있어야 하는지 생각해보세요~

ksundong commented on backend/src/main/java/com/codesquad/baseball/exceptions/oauth/InvalidJwtTokenException.java:
```diff
@@ -0,0 +1,14 @@
+package com.codesquad.baseball.exceptions.oauth;
+
+public class InvalidJwtTokenException extends RuntimeException {
+    public static final String BAD_TOKEN = "Authorization의 값이 잘못되었습니다! 형식은 bearer TOKEN_STRING 이어야 합니다";
+    public static final String EXPIRED_JWT_EXCEPTION = "만료된 토큰입니다";
+    public static final String UNSUPPORTED_JWT_EXCEPTION = "지원되지 않는 토큰입니다";
+    public static final String MALFORMED_JWT_EXCEPTION = "토큰 형식이 잘못되었습니다";
+    public static final String SIGNATURE_EXCEPTION = "토큰의 시그니쳐 검증을 실패했습니다";
+    public static final String ILLEGAL_ARGUMENT_EXCEPTION = "아규먼트 형식이 잘못되었습니다";
```
👍

ksundong commented on backend/src/main/java/com/codesquad/baseball/exceptions/oauth/InvalidRefreshTokenException.java:
```diff
@@ -0,0 +1,9 @@
+package com.codesquad.baseball.exceptions.oauth;
+
+public class InvalidRefreshTokenException extends InvalidTokenException {
+    public static final String INVALID_REFRESH_TOKEN = "사용자의 리프래시 토큰이 아닙니다";
```
유효하지 않은 리프레시 토큰이라는 내용이 더 정확하겠죠?

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/DataInitializer.java:
```diff
@@ -0,0 +1,68 @@
+
+    private void registerPlayers(Team team, HashMap<String, List<String>> testDataDTO) {
+        List<String> playerNames = testDataDTO.get(team.getTeamName());
+        IntStream.range(0, playerNames.size() - 1).forEach(playerIndex -> {
+            teamService.addHitter(team, playerNames.get(playerIndex), playerIndex + 1);
+        });
```
그냥 for문 쓰셔도 됩니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/GameService.java:
```diff
@@ -0,0 +1,148 @@
+        this.teamService = teamService;
+    }
+
+    @Transactional(readOnly = true)
```
클래스 레벨에 readonly = true를 붙여주면 해당 클래스의 모든 메서드에 readonly가 적용됩니다.
필요한 부분에 Transactional 애노테이션을 붙여주면 좀 더 깔끔한 코드를 유지할 수 있습니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/GameService.java:
```diff
@@ -0,0 +1,148 @@
+
+    @Transactional(readOnly = true)
+    public GamesDTO showGames() {
+        Iterable<Game> games = gameRepository.findAll();
+        List<GameDTO> gameDTOS = StreamSupport.stream(games.spliterator(), false)
+                .map(game -> {
+                    Team homeTeam = teamService.findTeam(game.homeTeamId());
+                    Team awayTeam = teamService.findTeam(game.awayTeamId());
+                    return GameDTO.from(game, homeTeam, awayTeam);
+                }).collect(Collectors.toList());
```
Stream의 활용 좋습니다.
정적 팩토리 메서드 네이밍 컨벤션만 지켜주시면 좋을 것 같아요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/GameService.java:
```diff
@@ -0,0 +1,148 @@
+    @Transactional
+    public void joinIn(int gameId, String userId) {
+        Game game = findGame(gameId);
+        if (game.isOccupied()) {
+            throw new GameAlreadyOccupiedException(gameId);
+        }
```
이건 joinGame에서 발생시켜줘도 되겠죠?

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/GameService.java:
```diff
@@ -0,0 +1,148 @@
+        gameRepository.deleteAll();
+    }
+
+    private PitchDTO createPitchDTO(Game game, PitchResult pitchResult) {
+        TeamDetailDTO homeTeam = teamDetailDTO(game, TeamType.HOME);
+        TeamDetailDTO awayTeam = teamDetailDTO(game, TeamType.AWAY);
+        GameStatusDTO gameStatusDTO = GameStatusDTO.from(game);
+        GameScoreDTO gameScoreDTO = GameScoreDTO.from(game);
+        PitchResultDTO pitchResultDTO = PitchResultDTO.from(pitchResult, game);
+
+        return new PitchDTO(homeTeam, awayTeam, gameStatusDTO, gameScoreDTO, pitchResultDTO);
+    }
+
```
사용되는 메서드 아래에 위치시켜주세요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/GameService.java:
```diff
@@ -0,0 +1,148 @@
+    private TeamDetailDTO teamDetailDTO(Game game, TeamType teamType) {
+        if (teamType == TeamType.HOME) {
+            return teamDetailDTO(game.homeTeam());
+        } else {
+            return teamDetailDTO(game.awayTeam());
+        }
```
else는 빼줘도 되겠죠?

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/GameService.java:
```diff
@@ -0,0 +1,148 @@
+        return TeamDetailDTO.from(team, players);
+    }
+
+    private List<PlayerDTO> playerDTOsByTeamParticipant(TeamParticipatingInGame participant) {
+        List<PlayerDTO> playerDTOS = new ArrayList<>();
+        for (PlayerParticipatingInGame playerParticipant : participant.getPlayers()) {
+            Player player = playerRepository.findById(playerParticipant.getPlayer())
+                    .orElseThrow(() -> new PlayerNotFoundException(PlayerNotFoundException.FIND_PLAYER_FAILED));
+            playerDTOS.add(PlayerDTO.from(player, playerParticipant));
+        }
+        return playerDTOS;
+    }
```
일단 메서드명이 불필요하게 길다는 생각이 들어요.
그리고 private 메서드로 표현될만한 메서드도 아니라는 생각이 듭니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/GameService.java:
```diff
@@ -0,0 +1,148 @@
+        return playerDTOS;
+    }
+
+    private void registerPlayersInGame(Team team, TeamParticipatingInGame teamParticipant) {
+        team.getPlayers().stream()
+                .filter(Player::isHitter)
+                .forEach(teamParticipant::addPlayer);
+        Player pitcher = team.getPlayers()
+                .stream()
+                .filter(Player::isPitcher)
+                .findFirst()
+                .orElseThrow(() -> new PlayerNotFoundException(PlayerNotFoundException.FIND_PITCHER_FAILED));
+        teamParticipant.addPlayer(pitcher, PitcherPosition.SP);
+    }
```
여기도 private 메서드가 아닌 편이 낫겠네요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/GoogleApiRequester.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public ReceiveAccessTokenDTO requestAccessToken(String authorizationCode) {
+        RequestAccessTokenDTO requestAccessTokenDTO = new RequestAccessTokenDTO(oAuthConfigManager.clientId(), oAuthConfigManager.clientSecret(), authorizationCode, REDIRECT_URI);
```
이런건 빌더 써주세요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/GoogleUrlService.java:
```diff
@@ -0,0 +1,30 @@
+    public static final String GOOGLE_SCOPE_PROFILE = "https://www.googleapis.com/auth/userinfo.profile";
+    public static final String REDIRECT_URI = "http://localhost:8080/users/callback";
+    //GOOGLE APIS
+    private static final String GOOGLE_API_GET_LOGIN_PAGE = "https://accounts.google.com/o/oauth2/v2/auth";
+    private final OAuthConfigManager oAuthConfigManager;
```
```suggestion
    private static final String GOOGLE_API_GET_LOGIN_PAGE = "https://accounts.google.com/o/oauth2/v2/auth";
    
    private final OAuthConfigManager oAuthConfigManager;
```

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/JwtBuilder.java:
```diff
@@ -0,0 +1,59 @@
+                .build();
+    }
+
+    private String buildAccessToken(Date now, String id) {
+        Claims claims = Jwts.claims();
+        claims.put(USER_ID, id);
+        return buildToken(now, id, ACCESS_TOKEN_EXPIRATION);
+    }
+
+    private String buildRefreshToken(Date now, String id) {
+        Claims claims = Jwts.claims();
+        claims.put(USER_ID, id);
+        return buildToken(now, id, REFERSH_TOKEN_EXPIRATION);
+    }
```
각각이 다른 곳에서 만들어지는게 낫겠죠?

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/JwtBuilder.java:
```diff
@@ -0,0 +1,59 @@
+        return buildToken(now, id, REFERSH_TOKEN_EXPIRATION);
+    }
+
+    private String buildToken(Date now, String id, Long expirationTime) {
+        Claims claims = Jwts.claims();
+        claims.put(USER_ID, id);
+        return Jwts.builder()
+                .signWith(SignatureAlgorithm.HS256, oAuthConfigManager.serverSecret())
+                .setClaims(claims)
+                .setIssuer(ISSUER)
+                .setIssuedAt(now)
+                .setExpiration(new Date(now.getTime() + expirationTime))
+                .compact();
+    }
```
토큰을 Build 해주는 것도 따로 만들어봅시다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/JwtVerifier.java:
```diff
@@ -0,0 +1,55 @@
+        return (String) claims.getBody().get(USER_ID);
+    }
+
+    public void verifyAccessToken(AccessTokenDTO accessTokenDTO) {
+        String userId = accessTokenDTO.getUserId();
+        User user = userService.findUserByUserId(userId);
+        user.verifyAccessToken(accessTokenDTO.getAccessToken());
+    }
+
+    public void verifyRefreshToken(RefreshTokenDTO refreshTokenDTO) {
+        String userId = refreshTokenDTO.getUserId();
+        User user = userService.findUserByUserId(userId);
+        user.verifyRefreshToken(refreshTokenDTO.getRefreshToken());
+    }
```
왜 user가 token을 verify할까요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/JwtVerifier.java:
```diff
@@ -0,0 +1,55 @@
+        user.verifyRefreshToken(refreshTokenDTO.getRefreshToken());
+    }
+
+    private Jws<Claims> verifyJwt(String jwt) {
+        try {
+            return Jwts.parser().setSigningKey(oAuthConfigManager.serverSecret()).parseClaimsJws(jwt);
+        } catch (ExpiredJwtException e) {
+            throw new InvalidJwtTokenException(InvalidJwtTokenException.EXPIRED_JWT_EXCEPTION);
+        } catch (UnsupportedJwtException e) {
+            throw new InvalidJwtTokenException(InvalidJwtTokenException.UNSUPPORTED_JWT_EXCEPTION);
+        } catch (MalformedJwtException e) {
+            throw new InvalidJwtTokenException(InvalidJwtTokenException.MALFORMED_JWT_EXCEPTION);
+        } catch (SignatureException e) {
+            throw new InvalidJwtTokenException(InvalidJwtTokenException.SIGNATURE_EXCEPTION);
+        } catch (IllegalArgumentException e) {
+            throw new InvalidJwtTokenException(InvalidJwtTokenException.ILLEGAL_ARGUMENT_EXCEPTION);
+        }
+    }
```
이 메서드도 사용되는 위치 아래에 위치시켜주세요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/OAuthConfigManager.java:
```diff
@@ -0,0 +1,51 @@
+
+        } catch (IOException e) {
+            logger.error(LOADING_OAUTH_CONFIG_FAILED_ERROR_MESSAGE);
+            e.printStackTrace();
```
이거 말고 logger로 찍어봅시다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/OAuthConfigManager.java:
```diff
@@ -0,0 +1,51 @@
+        } catch (IOException e) {
+            logger.error(LOADING_OAUTH_CONFIG_FAILED_ERROR_MESSAGE);
+            e.printStackTrace();
+            System.exit(1);
```
스프링 Application을 종료할 때는
https://stackoverflow.com/questions/22944144/programmatically-shut-down-spring-boot-application
이곳을 참고해서 종료시켜봅시다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/TeamService.java:
```diff
@@ -0,0 +1,60 @@
+        teamRepository.deleteAll();
+    }
+
+    private Player createPlayer(String playerName, int uniformNumber, PlayerRole playerRole) {
+        return new Player.Builder()
+                .playerName(playerName)
+                .uniformNumber(uniformNumber)
+                .role(playerRole)
+                .build();
+    }
```
이런 편의 메서드가 왜 Service에 있는지 모르겠어요.
Player가 가지면 안될까요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/OAuthConfigManager.java:
```diff
@@ -0,0 +1,51 @@
+    public ServerSecretDTO loadOAuthConfig() {
+        ServerSecretDTO serverSecretDTO = null;
+        try {
+            InputStream inputStream = OAuthConfigManager.class.getResourceAsStream("/server.secret.json");
```
try-with-resources 를 사용해주세요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/TestDataLoader.java:
```diff
@@ -0,0 +1,37 @@
+    public HashMap<String, List<String>> loadTestData() {
+        HashMap<String, List<String>> testData = null;
+        try {
+            InputStream inputStream = TestDataLoader.class.getResourceAsStream("/testData.json");
```
여기도 try-with-resources를 사용해주세요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/TestDataLoader.java:
```diff
@@ -0,0 +1,37 @@
+    private static final String LOADING_TEST_DATA_FAILED_ERROR_MESSAGE = "테스트 데이터 파일을 읽는 과정에서 문제가 발생했습니다. 프로그램을 종료합니다";
+    private static final String LOADING_TEST_DATA_COMPLETE = "loading oauth config complete!";
+
+    public HashMap<String, List<String>> loadTestData() {
+        HashMap<String, List<String>> testData = null;
```
왜 HashMap을 리턴해주시나요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/UserService.java:
```diff
@@ -0,0 +1,64 @@
+        return userRepository.save(new User(userInfoDTO, jwtTokenDTO));
+    }
+
+    public Optional<User> findOptionalUserByUserId(String userId) {
+        return userRepository.findByUserId(userId);
+    }
```
이런 메서드가 필요할까요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/UserService.java:
```diff
@@ -0,0 +1,64 @@
+            userRepository.save(user);
+            return;
+        }
+        createUser(userInfoDTO, jwtTokenDTO);
```
private에서 public 메서드를 사용하면, public 메서드가 overriding되면 동작이 바뀔 우려가 있습니다.

ksundong commented on backend/src/main/java/com/codesquad/baseball/service/UserService.java:
```diff
@@ -0,0 +1,64 @@
+        return newTokenDTO;
+    }
+
+    private void processUserData(UserInfoDTO userInfoDTO, Optional<User> optionalUser, JwtTokenDTO jwtTokenDTO) {
+        if (optionalUser.isPresent()) {
+            User user = optionalUser.get();
+            user.update(userInfoDTO);
+            user.updateAccessToken(jwtTokenDTO.getAccessToken());
+            user.updateRefreshToken(jwtTokenDTO.getRefreshToken());
+            userRepository.save(user);
+            return;
+        }
```
이곳 로직은 다시 구현해보시는게 좋을 것 같아요~

ksundong replied:
Optional로 받을 이유가 전혀 없어보이는 코드입니다.
차라리 null체크를 직접 해주는게 좋을 것 같네요.

ksundong commented on backend/src/main/java/com/codesquad/baseball/utils/ControllerUtil.java:
```diff
@@ -0,0 +1,20 @@
+            throw new ExtractUserIdFromTokenFailedException(ExtractUserIdFromTokenFailedException.EXTRACTION_FAILED);
+        }
+        return userId;
+    }
```
Spring에서 Custom Converter나 DataBinder를 구현하는 방법이 더 낫지 않을까요?

ksundong commented on backend/src/main/java/com/codesquad/baseball/utils/PitchRandomTable.java:
```diff
@@ -0,0 +1,41 @@
+
+public class PitchRandomTable {
+
+    private static final PlayType[] PITCH_RANDOM_TABLE = {
+            PlayType.STRIKE,
+            PlayType.STRIKE,
+            PlayType.STRIKE,
+            PlayType.STRIKE,
+            PlayType.BALL,
+            PlayType.BALL,
+            PlayType.BALL,
+            PlayType.BALL,
+            PlayType.BALL,
+            PlayType.HOMERUN,
+            PlayType.STRIKE,
+            PlayType.STRIKE,
+            PlayType.STRIKE,
+            PlayType.STRIKE,
+            PlayType.STRIKE,
+            PlayType.HITS,
+            PlayType.HITS,
+            PlayType.STRIKE,
+            PlayType.STRIKE,
+            PlayType.STRIKE,
+            PlayType.HOMERUN,
+            PlayType.STRIKE,
+            PlayType.STRIKE,
+            PlayType.HITS,
+            PlayType.STRIKE,
+    };
```
이렇게 많은 이유가 무엇인가요? 확률 때문인가요?

ksundong commented on backend/src/main/resources/application.properties:
```diff
@@ -0,0 +1,16 @@
+spring.datasource.initialization-mode=always
+spring.datasource.schema=classpath:schema.sql
+# encoding
+server.servlet.encoding.charset=UTF-8
+server.servlet.encoding.enabled=true
+server.servlet.encoding.force=true
```
인코딩 설정이 왜 필요했나요?

ksundong commented on backend/src/test/java/com/codesquad/baseball/domain/TeamRepositoryTest.java:
```diff
@@ -0,0 +1,92 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest
+@Transactional
+class TeamRepositoryTest {
```
Repository가 아닌 도메인 객체를 중점적으로 테스트 해보세요.

ksundong commented on backend/src/test/java/com/codesquad/baseball/domain/TeamRepositoryTest.java:
```diff
@@ -0,0 +1,92 @@
+    @Test
+    @DisplayName("팀에 플레이어를 추가할 수 있어야 함")
+    void testInsertPlayerIntoTeam() {
+        Team team = createTeam(testTeamNames[0]);
+        List<Player> players = createPlayers();
+        players.forEach(team::addPlayer);
+        teamRepository.save(team);
+        team = findTeamById(team.getId());
+        assertThat(team.numberOfPlayer()).isEqualTo(players.size());
+        logger.debug("team : {}", team);
```
무엇이 테스트를 준비하기 위한 코드고
무엇이 테스트 대상인 코드인지 명확하지 않네요.
그리고 log를 찍어서 확인할 필요는 없는 것 같아요.

ksundong commented on backend/src/test/java/com/codesquad/baseball/domain/TeamRepositoryTest.java:
```diff
@@ -0,0 +1,92 @@
+            "TEST-TEAM-C",
+    };
+    private static final String[] playerNames = {
+            "TEST-PLAYER-A",
+            "TEST-PLAYER-B",
+            "TEST-PLAYER-C",
```
Integeration 테스트라면 좀 더 이름같은 이름으로 바꿔보시는걸 추천드릴게요.

ksundong commented on backend/src/test/java/com/codesquad/baseball/domain/TeamRepositoryTest.java:
```diff
@@ -0,0 +1,92 @@
+    private List<Player> createPlayers() {
+        List<Player> players = new ArrayList<>();
+        for (int i = 0; i < playerNames.length; i++) {
+            Player player = createHitter(playerNames[i], i + 3);
```
i + 3의 의미는 무엇인가요?

ksundong commented on backend/src/test/java/com/codesquad/baseball/domain/TeamRepositoryTest.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    private void testCreateTeam(Team team, String teamName) {
+        logger.debug("saved team : {}", team);
```
테스트 코드에서 로깅하는건 별 의미가 없다고 생각해요~

ksundong commented on backend/src/test/java/com/codesquad/baseball/domain/GameRepositoryTest.java:
```diff
@@ -0,0 +1,638 @@
+    @Autowired
+    private TeamRepository teamRepository;
+
+    @Test
+    @DisplayName("게임을 생성할 수 있어야 합니다.")
+    void testCreateGame() {
+        Game game = createGame(GAME_TITLE);
+        assertThat(game).isNotNull();
+    }
```
이것은 굳이 Spring Container가 없이도, 테스트 할 수 있는 단위테스트죠?
분리해봅시다.

ksundong commented on backend/src/test/java/com/codesquad/baseball/domain/GameRepositoryTest.java:
```diff
@@ -0,0 +1,638 @@
+
+    @Test
+    @DisplayName("피치의 결과가 볼인 상황을 테스트합니다")
+    void testPitchIsBall() {
```
test라는 prefix는 굳이 없어도 되지 않을까요?

ksundong commented on backend/src/test/java/com/codesquad/baseball/domain/GameRepositoryTest.java:
```diff
@@ -0,0 +1,638 @@
+    @Test
+    @DisplayName("피치의 결과가 안타인 상황을 테스트합니다")
+    void testPitchIsHits() {
+        Game game = createGame(GAME_TITLE);
+        //1안타후엔 첫번째 타자가 1루에 가있어야 함
+        int firstHitter = game.currentHitter();
+        doHits(game);
+        assertThat(game.firstBaseRunner()).isEqualTo(firstHitter);
+        //2안타후엔 첫번째 타자가 2루, 두번째가 1루
+        int secondHitter = game.currentHitter();
+        doHits(game);
+        assertThat(game.firstBaseRunner()).isEqualTo(secondHitter);
+        assertThat(game.secondBaseRunner()).isEqualTo(firstHitter);
+        //3안타후엔 첫번째 타자가 3루, 두번째가 2루 세번째는 3루
+        int thirdHitter = game.currentHitter();
+        doHits(game);
+        assertThat(game.firstBaseRunner()).isEqualTo(thirdHitter);
+        assertThat(game.secondBaseRunner()).isEqualTo(secondHitter);
+        assertThat(game.thirdBaseRunner()).isEqualTo(firstHitter);
+        //4안타후엔 첫번째 타자가 백홈, 두번째가 3루 세번째는 2루 네번째가 1루, 원정팀 1점
+        int fourthHitter = game.currentHitter();
+        PitchResult backHomeResult = doHits(game);
+        assertThat(game.firstBaseRunner()).isEqualTo(fourthHitter);
+        assertThat(game.secondBaseRunner()).isEqualTo(thirdHitter);
+        assertThat(game.thirdBaseRunner()).isEqualTo(secondHitter);
+        assertThat(backHomeResult.numberOfRunners()).isEqualTo(1);
+        assertThat(backHomeResult.findRunnerByOrder(0)).isEqualTo(firstHitter);
+        assertThat(game.awayTeamScore()).isEqualTo(1);
+        //공수교대 후 5안타하면 홈팀 2점
+        threeOut(game);
+        IntStream.range(0, 5).forEach(value -> game.pitch(PlayType.HITS));
+        assertThat(game.homeTeamScore()).isEqualTo(2);
+        //공수교대 후 5안타하면 원정팀 3점, 홈팀 여전히 2점
+        threeOut(game);
+        IntStream.range(0, 5).forEach(value -> game.pitch(PlayType.HITS));
+        assertThat(game.awayTeamScore()).isEqualTo(3);
+        assertThat(game.homeTeamScore()).isEqualTo(2);
```
각각의 테스트를 분리해봅시다.

ksundong commented on backend/src/test/java/com/codesquad/baseball/domain/GameRepositoryTest.java:
```diff
@@ -0,0 +1,638 @@
+        assertThat(pitchResult.getGameState()).isEqualTo(GameState.GAME_OVER);
+    }
+
+    private void testHistoryOfCurrentInning(Game game) {
```
test코드가 너무 장황하죠?
좋은 테스트 코드는 한 눈에 알아볼 수 있어야 합니다.

ksundong reviewed (changes requested):
안녕하세요! NAS! 리뷰어 Dion입니다.
야구게임은 상당히 어려웠던 프로젝트로 기억하고 있습니다. 어떤 기술을 선택할지도 되게 고민이 많이 되었던 것 같아요.
그래서 아마 제가 드리는 피드백이 어쩌면 너무 과할 수도 있다는 생각이 들어요.

이번 피드백은 전부 반영해주지 않으셔도 좋습니다. 왜냐하면 2주라는 기간이 짧기 때문에 아마 다 못하셨을거라고 생각되거든요.
제가 추천드리는 방법은, 일단 받은 피드백을 목록을 만드시고, 정리하신 다음 다음 미션에서 좀 더 개선된 형태로 진행하시는걸 추천드립니다.
되도록이면 매 주마다 PR을 보내셔서 지속적으로 피드백 받으시는걸 추천드려요.

## 피드백

- 클래스의 라인 수는 되도록이면 100줄 내외가 될 수 있도록 작성해보세요.
- Integeration Test와 Unit Test를 구분지어보시고, Unit Test를 중심으로 생각해봅시다.
- private 메서드의 개수보다는 뎁스가 3개 이상으로 늘어나지 않도록 클래스를 분리해봅시다. (최대한 2개까지)
- 코드는 코드를 읽는 사람과의 소통입니다. 처음 코드를 읽는 사람이 편할 수 있도록 코드를 작성해봅시다.
- 빌더 패턴은 만병통치약이 아닙니다. 필드가 많아진다면, 분리를 하는 것이 객체지향적인 사고라고 볼 수 있습니다.
- 예외 처리에 좀 더 신경써주세요.
- `HttpServletRequest`를 최대한 Controller에서 사용하지 않게끔 코딩해보세요.
- switch case문, else문을 최대한 피하는 방식으로 코딩해보세요.
- 코드블럭의 depth는 곧 그 코드블럭의 복잡도를 의미합니다. 뎁스를 최소화하는 방향으로 코딩해주세요.
- 도메인이 DTO를 아는 방식의 설계는 최대한 피해주시고, Converter 역할을 하는 DTO정도만 허용을 하는식으로 Domain이 DTO에 영향받지 않게끔 코딩해주세요.
- Builder 정도는 Lombok 사용을 추천드립니다. 다만 Builder가 남용되면 그건 그것대로 곤란합니다.
- 자바 컨벤션은 이제 리뷰 받으면 안됩니다. 습관화 해주세요!
- Optional의 사용이 필요한 경우와 그냥 null을 사용해도 되는 경우를 구분해보세요.
- IntStreamRange를 사용하는 것보다 for문을 사용하는게 더 나을 수도 있습니다.

2주간 프로젝트 수행하느라 고생많으셨고, 주말에 푹 쉬세요~!
궁금하시거나, 도움이 필요한 부분 이해가 가지 않는 부분등은 코멘트 남겨주시면 됩니다!

ksundong commented:
> # 야구게임 구현
> * 야구는 테스트코드를 작성해가면서 기능구현을 했습니다
> * 그렇긴 한데, 아직 테스트코드를 작성하는게 많이 미숙한 듯 합니다
>   ![image](https://user-images.githubusercontent.com/41663269/118095911-a91fb700-b40b-11eb-93c2-5e15b2f3d13f.png)

제가 요청드린 테스트는 단위 테스트입니다. 아직 단위 테스트와 통합 테스트의 구분이 안되어 계신것 같아요.
단위테스트는 단위테스트 대로, 통합테스트는 통합테스트 대로 테스트 클래스를 구분지어주세요.
단위 테스트에 대해서 주말에 한 번 공부해보시면 좋을 것 같네요.

> ### 어그리게이트
> * Game과 Team을 각각 분리된 어그리게이트로 만들었습니다.
> * Game을 어그리게이트 루트로 삼아서 게임에 관련된 모든 정보를 관리하도록 했습니다
> * Game과 별개로 팀 자체의 정보는 Team을 어그리게이트 루트로 삼아서 관리하도록 설계했습니다.
> * Game을 만들게되면 Team의 정보를 이용해서 팀이 게임에 참여했다는 정보를 저장하도록 했습니다.(TeamParticipatingInGame)
> * 이렇게 해서 게임마다 선수의 타순을 자유롭게 설정할 수 있도록 설계했습니다.

aggregate가 로직을 한 클래스(aggreate root)에 몰아두자가 아니라, endpoint를 root로 두자라고 생각하시면 좋을 것 같습니다.
각각의 aggregate는 여러 객체의 협력 구조로 이루어지는게 좋을 것 같습니다.

> ### 토큰 리프래시
> * 액세스토큰이 만료되면 리프래시 토큰으로 토큰 리프래시 요청을 해서 새로운 액세스 토큰과 리프래시 토큰을 받아오도록 구현했습니다
>   ![image](https://user-images.githubusercontent.com/41663269/118113903-d5dec900-b421-11eb-96b0-27ecea3c99e8.png)

리프레시 토큰은 JWT로 구현해주셔도 괜찮지만, 딱히 JWT를 쓸 이유도 없다고 생각합니다.
차라리 민감한 정보를 담아둘 수 있는 opaque token이라던가, cache server를 이용하는 방법도 있겠죠.
구현 방법은 여러가지지만, 나쁘지 않은 구현이라고 생각해요!

제가 회사에서 구현한 token 관련 로직은 access token을 요청할 때, access Token과 refresh Token을 발급해주고, refresh Token의 만료기간이 가까워 온다면, 만료시키고 새로운 refresh Token을 발급해주는 방식으로 구현했었습니다.

> ### 토큰 검사
> * 토큰은 인터셉터를 이용해서 검사하도록 구현했습니다.
> * `@Auth`어노테이션 유뮤를 가지고 검사할지 말지를 결정할 수 있도록 구현했습니다
> 
> ```java
> @Auth
> @PatchMapping("/{gameId}")
> public void joinIn(@PathVariable("gameId") int gameId, HttpServletRequest request) {/*...*/}
> ```

인터셉터가 적용될 path를 지정해주는 편이 좋지 않았을까 싶네요.


고생하셨습니다~

Malloc72P commented:
오우 디온... 정말 세심한 리뷰 감사드려요!

오늘안에 다 반영해보려고 했는데 힘들 것 같네요

내일 이어서 더 해보겠습니다

Malloc72P committed:
docs : 애노테이션을 설명하는 주석 추가

Malloc72P committed:
refactor : 인터셉터에 서비스 주입방식 변경

Malloc72P committed:
refactor : 예외를 상위클래스로 받아서 처리하도록 수정

Malloc72P committed:
refactor : JwtToken예외도 묶어서 처리할 수 있도록 수정

Malloc72P committed:
fix : 잘못된 메서드명 수정

Malloc72P committed:
refactor : path pattern을 잘 부여해서 원하는 메서드만 등록되도룩 수정

Malloc72P committed:
refactor : 리다이렉션은 304로 하도록 수정

Malloc72P committed:
feat : 스네이크 케이스 회피

Malloc72P committed:
refactor : 너무 깊은 depth 해소

Malloc72P committed:
refactor : ArgumentResolver 적용

- ArgumentResolver를 적용해서 컨트롤러의 핸들러 메서드에서 HttpServletRequest를 받아서 값을 꺼내는 일이 없도록 수정함

Malloc72P commented:
주말내에 리뷰해주신 내용을 다 공부해보고 반영해보려고 했는데요, 생각처럼 잘 안되네요 ㅠㅠ

다음주가 되면 새 프로젝트가 시작돼서 많이 바쁘겠지만, 짬짬히 시간내서 공부하고 반영할께요!

ksundong commented on backend/src/main/java/com/codesquad/baseball/config/interceptor/RefreshInterceptor.java:
```diff
@@ -11,13 +13,17 @@
     @Resource
     private JwtVerifier jwtVerifier;
 
+    @Resource
```
리소스 애노테이션도 그리 좋은 선택은 아닙니다.
`@Autowired`랑 다를게 크게 없어요.

Malloc72P replied:
위에서 말씀해주신 것처럼 스프링 공부를 더해보고 생성자주입으로 넣을 수 있도록 고쳐보겠습니다!

ksundong commented on backend/src/main/java/com/codesquad/baseball/config/interceptor/RefreshInterceptor.java:
```diff
@@ -39,7 +45,8 @@ public void afterCompletion(HttpServletRequest request, HttpServletResponse resp
         RefreshTokenDTO refreshTokenDTO = verifyRefreshToken(request);
         jwtVerifier.verifyRefreshToken(refreshTokenDTO);
-        request.setAttribute(USER_ID_KEY, refreshTokenDTO.getUserId());
+        User certifiedUser = userService.findUserByUserId(refreshTokenDTO.getUserId());
+        request.setAttribute(CERTIFIED_USER, certifiedUser);
```
자, 잘 생각해봅시다. User를 리턴해주면, 우리가 지금까지 설정했던 관계들이 전부 꼬이게 되지 않나요?
Controller가 도메인을 알게될겁니다. 별로 좋지 못하겠죠?

Malloc72P replied:
그러네요... 인터셉터도 그렇고, 디온이 말씀해주신대로 컨트롤러에서 유저를 알게되버리니까 의도치않은 의존관계가 생겨버리는군요... 기존에 하던대로 토큰에서 userId만 추출해서 넘겨주는 편이 더 나았겠네요

ksundong reviewed (approved):
마지막 리뷰입니다.
주말에 피드백 적용하느라 고생하셨어요~
